go/oasis-node/cmd/common/metrics/metrics.go

```diff
@@ -179,56 +313,533 @@ func (s *pushService) worker() {
+	}
+
+	d, err := NewDiskService()
+	if err != nil {
+		return nil, err
+	}
+
+	m, err := NewMemService()
+	if err != nil {
+		return nil, err
+	}
+
+	c, err := NewCPUService()
+	if err != nil {
+		return nil, err
+	}
+
+	n, err := NewNetService()
+	if err != nil {
+		return nil, err
+	}
+
 	return &pushService{
 		BaseBackgroundService: svc,
-		pusher:                push.New(addr, jobName).Grouping("instance", instanceLabel),
+		pusher:                pusher,
 		interval:              interval,
+		dService:              d.(*diskService),
+		mService:              m.(*memService),
+		cService:              c.(*cpuService),
+		nService:              n.(*netService),
 	}, nil
 }
 
 // New constructs a new metrics service.
 func New(ctx context.Context) (service.BackgroundService, error) {
-	mode := viper.GetString(cfgMetricsMode)
+	mode := viper.GetString(CfgMetricsMode)
 	switch strings.ToLower(mode) {
-	case metricsModeNone:
+	case MetricsModeNone:
 		return newStubService()
-	case metricsModePull:
+	case MetricsModePull:
 		return newPullService(ctx)
-	case metricsModePush:
+	case MetricsModePush:
 		return newPushService()
 	default:
 		return nil, fmt.Errorf("metrics: unsupported mode: '%v'", mode)
 	}
 }
 
+type diskService struct {
+	service.BaseBackgroundService
+	sync.Mutex
+
+	dataDir string
+	// TODO: Should we monitor I/O of children PIDs as well?
+	pid      int
+	interval time.Duration
+
+	diskUsageGauge          prometheus.Gauge
+	diskIOReadBytesGauge    prometheus.Gauge
+	diskIOWrittenBytesGauge prometheus.Gauge
+}
+
+func (d *diskService) Start() error {
+	go d.worker()
+	return nil
+}
+
+// updateDiskUsage walks through dataDir and updates diskUsageGauge attribute.
+func (d *diskService) updateDiskUsage() error {
+	d.Lock()
+	defer d.Unlock()
+
+	// Compute disk usage of datadir.
+	var duBytes int64
+	err := filepath.Walk(d.dataDir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return errors.Wrap(err, fmt.Sprintf("disk usage metric: failed to access file %s", path))
+		}
+		duBytes += info.Size()
+		return nil
+	})
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("disk usage metric: failed to walk directory %s", d.dataDir))
+	}
+	d.diskUsageGauge.Set(float64(duBytes))
+
+	return nil
+}
+
+// updateDiskUsage reads process info and updates diskIO{Read|Written}BytesGauge attributes.
+func (d *diskService) updateIO() error {
+	d.Lock()
+	defer d.Unlock()
+
+	// Obtain process I/O info.
+	proc, err := procfs.NewProc(d.pid)
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("disk I/O metric: failed to obtain proc object for PID %d", d.pid))
+	}
+	procIO, err := proc.IO()
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("disk I/O metric: failed to obtain procIO object %d", d.pid))
+	}
+
+	d.diskIOWrittenBytesGauge.Set(float64(procIO.ReadBytes))
+	d.diskIOReadBytesGauge.Set(float64(procIO.WriteBytes))
+
+	return nil
+}
+
+func (d *diskService) worker() {
```

Comment:
It seems each service has its own worker that looks the same -- reporting on an interval. It would be better if services had an `Update` method which would be called by a common worker for all services.